domain/stub/service.go

```diff
@@ -186,3 +199,67 @@ func (s *StubService) GetExecSecretToken(ctx context.Context) (string, error) {
 
 	return provider.GetSecretToken(ctx, k8sprovider.ExecRBACResourceName)
 }
+
+// AgentBinaryFilter is a function that filters agent binaries based on the
+// given parameters. It returns a list of agent binaries that match the filter
+// criteria.
+type AgentBinaryFilter func(
+	ctx context.Context,
+	ss envtools.SimplestreamsFetcher,
+	env environs.BootstrapEnviron,
+	majorVersion,
+	minorVersion int,
+	streams []string,
+	filter coretools.Filter,
+) (coretools.List, error)
+
+// PreferredSimpleStreams is a function that returns the preferred streams
+// for the given version and stream.
+type PreferredSimpleStreams func(
+	vers *semversion.Number,
+	forceDevel bool,
+	stream string,
+) []string
+
+// EnvironAgentBinariesFinderFunc is a function that can be used to find agent binaries
+// from the simplestreams data sources.
+type EnvironAgentBinariesFinderFunc func(
+	ctx context.Context,
+	major,
+	minor int,
+	version semversion.Number,
+	requestedStream string,
+	filter coretools.Filter,
+) (coretools.List, error)
+
+// GetEnvironAgentBinariesFinder returns a function that can be used to find
+// agent binaries from the simplestreams data sources.
+func (s *StubService) GetEnvironAgentBinariesFinder(
+	agentBinaryFilter AgentBinaryFilter,
```

Comment:
I understand why the dependencies have been asked for in this way to help support testing and not to produce external changes.

I think however it shouldn't be on the caller to supply this information. Specifically we are trying to make the caller have to have less transitive dependencies and understanding of how the cake is made.

I think if we want to keep this pattern for testing purposes we should do what we do in the workers where this package provides the implementations of the dependencies that can be supplied to the constructor of the stub service.

It allows us to control the dependencies still but not bleed this out to callers.